### ISSUE TITLE
Memory leak in strided, valid-padded convolution with dynamic input shape

### ISSUE THREAD

**Issue by tazr** [NVIDIA/TensorRT#428]
## Description

GPU memory footprint of program keeps growing when running a strided, valid-padded convolution with dynamic input shape.

## Environment

**TensorRT Version**: 7.0.0.11
**GPU Type**: Titan X
**Nvidia Driver Version**: 436.30
**CUDA Version**: 10.0
**CUDNN Version**: 7.6.3.30
**Operating System + Version**: Windows 10 1903
**Python Version (if applicable)**: 3.7.6
**TensorFlow Version (if applicable)**: 1.14.0
**PyTorch Version (if applicable)**: N/A
**Baremetal or Container (if container which image + tag)**: N/A

## Steps To Reproduce

1. Create a TF model
```python
import tensorflow as tf
input = tf.placeholder(tf.float32, [1, None, None, 1], name='x')
layer = tf.nn.conv2d(input, tf.zeros((2, 2, 1, 1)), (1, 2, 2, 1), 'VALID')
output = tf.reduce_sum(layer, axis=[1, 2, 3], name='y')
with tf.Session() as sess:
  g = tf.get_default_graph().as_graph_def()
  tf.train.write_graph(g, ".", "leak.pb", as_text=False)
```

2. Convert to ONNX
```
python -m tf2onnx.convert --input leak.pb --output leak.onnx --inputs x:0 --outputs y:0 --opset 11
```

3. Build engine
```
trtexec --onnx=leak.onnx --saveEngine=leak.trt --explicitBatch --minShapes='x:0':1x2x2x1 --optShapes='x:0':1x128x128x1 --maxShapes='x:0':1x256x256x1
```

4. Run inference
```cpp
# include <array>
# include <fstream>
# include <random>
# include <cuda_runtime_api.h>
# include <NvInfer.h>
# include <samples/common/buffers.h>


class MyLogger : public nvinfer1::ILogger
{
  void log(Severity severity, char const * msg) override
  {
    std::cout << msg << std::endl;
  }
} logger;


int main()
{
  std::ifstream input("leak.trt", std::ios::binary);
  std::vector<unsigned char> model(std::istreambuf_iterator<char>(input), {});
  auto runtime = nvinfer1::createInferRuntime(logger);
  auto engine = runtime->deserializeCudaEngine(model.data(), model.size());
  auto context = engine->createExecutionContext();
  auto x_idx = engine->getBindingIndex("x:0");
  auto y_idx = engine->getBindingIndex("y:0");
  samplesCommon::ManagedBuffer x{ nvinfer1::DataType::kFLOAT, nvinfer1::DataType::kFLOAT };
  samplesCommon::HostBuffer y_host(1, nvinfer1::DataType::kFLOAT);
  samplesCommon::DeviceBuffer y_device(1, nvinfer1::DataType::kFLOAT);
  std:random_device rd;
  std::mt19937 gen(rd());
  std::uniform_int_distribution<> dis(1, 30);
  std::array<void *, 2> buffer;
  while (true)
  {
    auto x_shape = nvinfer1::Dims4(1, dis(gen) * 2, dis(gen) * 2, 1);
    x.hostBuffer.resize(x_shape);
    std::fill_n(static_cast<float*>(x.hostBuffer.data()), samplesCommon::volume(x_shape), 0.f);
    x.deviceBuffer.resize(x_shape);
    cudaMemcpy(x.deviceBuffer.data(), x.hostBuffer.data(), x.hostBuffer.nbBytes(), cudaMemcpyHostToDevice);
    context->setBindingDimensions(x_idx, x_shape);
    buffer[x_idx] = x.deviceBuffer.data();
    buffer[y_idx] = y_device.data();
    context->executeV2(buffer.data());
  }
  return 0;
}
```

This results in a slow but steady increase of GPU memory.

![Taskmgr_RvBVqLbcRX](https://user-images.githubusercontent.com/29096027/76466580-42a69f80-63e8-11ea-9e48-d1fe3a597b40.png)

Note that I don't have a leak no more
* if the convolution padding is changed to `'SAME'`,
* or if the stride is changed to `(1, 1, 1, 1)`.

For the reasons above, it is possible that this issue is different from  #351 , which has been fixed, since #351 is demonstrated on a non-strided, padded convolution.

**Comment by rmccorm4** [comment]
Hi @tazr,

Thanks for the detailed repro, it's much appreciated.

Unfortunately I couldn't repro the memory leak on either of my Ubuntu machines with V100 or P4 GPUs, and I don't have a Windows machine to test on. 

I passed this issue along to the TensorRT team to see if they can repro on Windows. We'll have to wait for them to look into it for now.

**Comment by tazr** [comment]
Thanks. That's unfortunate. Please tell me if you need anything else from my side to reproduce the issue. Note that the leak in my example is small and takes time to build up. You get a larger, faster leak when stacking those layers.

**Comment by tazr** [comment]
Since my driver was a bit old, I installed the latest one for my device (445.75) to rule out driver bugs solved since. I still have the leak with this driver.

**Comment by tazr** [comment]
@rmccorm4 Has anyone been able to reproduce this bug yet?

**Comment by rmccorm4** [comment]
Hi @tazr ,

Looks like this has been fixed in the next release. Please stay tuned.

**Comment by tazr** [comment]
Hey @rmccorm4, this issue is NOT fixed in 7.1.3.4. I also updated

* my driver to 451.48
* CUDA to 11.0.182
* CUDNN to 8.0.1.13

I still observe the slow but steady increase in GPU memory, just like before.

![Taskmgr_CaVcyJbD7V](https://user-images.githubusercontent.com/29096027/86451121-bfa0b080-bd1a-11ea-8325-3cbc37395be9.png)

Could you please reopen the issue and make sure it is at least reproduced on your side?

**Comment by tazr** [comment]
Sorry, I must have messed up. When I redo all the steps from the beginning, it works. Good job! You can please close the issue.
